fabfile.py

```diff
@@ -1,197 +1,518 @@
+    if not os.path.isfile(PRIVATE_SETTINGS):
+        raise ValueError(
+            "unable to find additional settings at {}".format(
+                PRIVATE_SETTINGS
+            )
+        )
+
+    with open(PRIVATE_SETTINGS) as privado:
+        result = json.load(privado)
+        err_template = "we require {} in your private settings"
+        if "db_password" not in result:
+            raise ValueError(err_template.format("a db password"))
+        if "additional_settings" not in result:
+            raise ValueError(err_template.format(
+                "an additional_settings dict (even if its empty)"
             ))
-        for db_command in db_commands(username):
-            local("heroku run --app {0} {1}".format(name, db_command))
+        if "proxy" not in result:
+            raise ValueError(err_template.format("proxy variable"))
+        if "host_string" not in result:
+            e = "we host string to be set, this should be 127.0.0.1 on test, \
+or the address you want to sync to on prod in your private settings"
+            raise ValueError(e)
+
+    return result
 
 
 @task
-def production_deploy():
-    # TODO include the supervisor commands
-    # synch the nginx conf
-    with lcd(fabfile_dir):
-        local("python manage.py migrate")
-        local("python manage.py collectstatic --noinput")
+def clone_branch(branch_name):
+    branch_env = Env(branch_name)
+    local(
+        "git clone -b {0} {1} {2}".format(
+            branch_name,
+            GIT_URL,
+            branch_env.project_directory
+        )
+    )
 
 
 @task
-def checkout_project():
-    checkout(get_requirements())
+def create_private_settings():
+    with open(PRIVATE_SETTINGS, "w") as privado:
+        json.dump(
+            dict(
+                proxy="",
+                db_password="",
+                host_string="",
+                additional_settings={}
+            ),
+            privado,
+            indent=4
+        )
 
 
 @task
-def create_db(username):
-    with lcd(fabfile_dir):
-        with warn_only():
-            # heroku somtimes has memory issues doing migrate
-            # it seems to work fine if we just migrate opal first
-            # it will later fail because content types haven't
-            # been migrated, but that's fine we'll do that later
-            local("python manage.py migrate opal")
-
-        for db_command in db_commands(username):
-            local(db_command)
+def create_empty_env(new_branch):
+    _deploy(new_branch)
+
+
+def _deploy(new_branch, backup_name=None):
+    if backup_name and not os.path.isfile(backup_name):
+        raise ValueError("unable to find backup {}".format(backup_name))
+
+    # the new env that is going to be live
+    new_env = Env(new_branch)
+
+    # the private settings
+    private_settings = get_private_settings()
+    env.host_string = private_settings["host_string"]
+
+    # Setup environment
+    pip_create_virtual_env(new_env)
```

Comment:
I think I want print to stdout as we perform each of these tasks somehow - either here or inside the functions...